openedx/openedx-platform
# MIT CCX: Use CCX Keys

This pull request implements the use of ccx-specific opaque keys (`CCXLocator` and `CCXBlockUsageLocator` throughout the codebase.  
### History

The request to implement custom opaque keys for ccx was first brought up in an architecture review of the PR referenced above.  It was felt that using such keys would allow ccx to better integrate with existing logging and analytics systems.  The outcome of this conversation was captured in jazkarta/edx-platform#43 

A separate PR (https://github.com/edx/edx-platform/pull/8259) has been opened to add the dependency upon the [custom key implementation](https://github.com/jazkarta/ccx-keys).  That PR is under review and should be merged prior to this.

During implementation of this code, @nedbat, @cpennington, @ormsbee, and @carsongee have all been pulled into discussions at various points.

The work has been overseen by @pdpinch and the team at MIT.
### Purpose

By creating a custom opaque-key implementation, the mechanism by which a user views CCX content (as opposed to the course from which the ccx was created) is simplified, and made to fall in line with existing mechanisms for retrieving course modules, assets and so on.  In addition, because there is a "locator" for a ccx that is distinct from that for the course, logging and analytics can register the use of CCX content separately from the use of the same materials for the original course.
### Impacts

This implementation adds a wrapper around the modulestore returned by `xmodule.modulestore.django.modulestore`.  When the CCX feature is enabled in settings, then wherever that method is called, a wrapped version of the modulestore will be returned.  This wrapper strips the ccx information out of locators and keys as queries are sent in to the modulestore and restores the information to the data that is returned.  Because of this, there is little in the system that is not impacted by this change.  For queries to the modulestore that use non-ccx keys or locations, this wrapper should be transparent, but it is present.

An additional implication of this work is that CCXs can not be created from legacy courses that use the pre-split deprecated form of course keys and locations. This restriction can be worked around by creating a re-run of such a course, which will convert it to the new form of storage and keys/locations.  It is therefore not considered a blocker.

In all other respects, the work should not have any direct impact on users.  CCX functionality continues to work as it did before this change.  There are no user-facing UI elements that have changed.

This work is integrated with previous CCX work, so it will be released initially only on edge, where CCX is already in use. 
### Manual Testing

In order to test these changes, you must begin with a course that uses the split-modulestore and has opaque-key course keys and locations (`course-v1:org+course+run...`).  Once that is in place, then you can follow these steps:
1.  Log in to the LMS an admin for the course, view the "instructor" tab and select "membership"
2.  In the course staff section, select the 'CCX Coach' role and add a user as a ccx coach.
3.  Log in as the CCX Coach user and visit the CCX Coach tab.
4.  Fill in a name for your ccx and submit to create it. You should be redirected to the Coach Dashboard view for the CCX.
5.  verify that when viewing the coach tab, you see a url that contains `ccx-v1:` and has `+ccx@n` in the identifier (`n` is the integer primary key of the CCX that was created). 
6.  verify that when viewing any other top level tab, you see only the course id (it should start with `course-v1:` and _not_ contain any mention of ccx).
7.  From the coach tab schedule some course content, but not all the items from the original course (so you can easily tell which you are seeing).
8.  From the coach tab, enroll a user in your ccx.
9.  Log in to the LMS as the student just enrolled by the coach.  You should see both the original course and the CCX on your dashboard.  They should be visually distinct
10.  Click on the CCX to go to the course info page.  It should look identical to the original course info page.  The URL for the page should contain `ccx-v1:` and `+ccx@n`.  (this verifies that you are viewing the ccx).
11.  Click on the courseware tab, verify that you see _only_ those course elements that were scheduled by the coach, and not any others.
12.  View and interact with the courseware.  Verify that everything works as expected.
13.  Check your progress report and verify that it operates as expected.
14.  Verify that the student cannot see the discussion tab (it should be hidden for any ccx).
15.  Log in as Coach and check the grade reports under `student admin` to verify that grades are being correctly calculated.  Update the grading policy and verify that it is applied to both the Coach grade views and the student progress view.
### Concerns

The modulestore wrapper is not well covered with tests.  We are not comfortable that we understand how best to approach the testing of this element.  We would appreciate any pointers in that regard.  

The management of CCX keys and locations versus Course keys and locations is a bit tricky and there are many possible edge cases we have yet to uncover.  There are a few places where manually unwrapping a ccx key is required to get course access information directly.  We would appreciate attention paid to this aspect.  If there are other places we have not yet found or thought of that might need special attention, please let us know.

It should be noted that at the time of this PR, there are 5 test failures and one error happening in my local development environment.  All of these are in the tests for the sysadmin dashboard and have to do with importing XML course content from git.  I believe the problem to be in my development environment, but if anyone (like @carsongee) has suggestions about fixing them should they also appear in jenkins, I'd appreciate it.
